convert2rhel/subscription.py

```diff
@@ -277,28 +287,131 @@ def args(self):
+        connection_opts = {}
+
+        url_parts = None
+        if self.server_url:
+            try:
+                url_parts = parse_url(self.server_url)
+            except ValueError:
+                # url_parts will remain None which will mean we return an empty dbus.DIctionary
+                pass
+
+        if url_parts:
+            if url_parts.hostname:
+                connection_opts["host"] = url_parts.hostname
+            if url_parts.port is not None:
+                connection_opts["port"] = url_parts.port
+
+            # Current subscription-manager server-side does not take query and fragments at this
+            # path but include them for future proofing.
+            path = urllib.parse.urlunsplit(("", "", url_parts.path, url_parts.query, url_parts.fragment))
+            if path:
+                connection_opts["handler"] = path
+
+        connection_opts = dbus.Dictionary(connection_opts, signature="ss")
+        return connection_opts
+
+    def __call__(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS-1.2.24 is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus-1.10.24 is installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8 & RHEL9
+                * dbus-1.12.x is installed and run by default
+                * python3-dbus-1.2.x is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html
+        """
+        # DBus demands that the full signature of a dictionary (what the datatypes of the keys and
+        # values will be) is known ahead of time.  An empty python dict doesn't have enough
+        # information and dbus-python doesn't like {"": ""} even though /bin/dbus-send thinks that
+        # is okay.  Use this instead.
+        EMPTY_DBUS_DICT = dbus.Dictionary({}, signature="ss")
+
+        loggerinst.debug("Getting a handle to the system dbus")
+        system_bus = dbus.SystemBus()
+
+        # Create a new bus so we can talk to rhsm privately (For security:
+        # talking on the system bus might be eavesdropped in certain scenarios)
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start(
+            i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer"
+        )
```

Comment:
One question that strikes me is, this piece of code can't fail when we run it? I'm not sure about the dbus-library, but that could rise an unexpected exception?

Reply:
This code can fail.  You are probably asking why it isn't part of the `try: finally:` that starts on the following line?  The reason for that is we only need the `try: finally:` to ensure we tear down the private DBus server that we're creating in this line.  If this line fails, then there won't be a registration server to tear down.

I'm assuming that register_server.Start won't fail after it creates the private DBus server and before it returns to us.  If it does fail there, we could end up with the server still running.  If that's the case, we'd also have to make the `finally:` clause handle both the server having been started and the case where it had not yet been started.

The current subscription-manager starts the server, logs it, and then returns so it seems pretty safe to let subscription-manager's code worry about this cornercase.

Reply:
Oh, I see... It totally makes sense now, and you're correct, my question was regarding the `try: finally:` block. In my point of view, I would expect that the server start would fail and start a rollback immediately (or worse, raise an exception and prevent  us to do a rollback)